### ISSUE TITLE
Indentation is weird with unicode symbols as operators, sometimes

### ISSUE THREAD

**Issue by wilbowma** [greghendershott/racket-mode#262]
I run into this a lot when writing Redex models. Here is an example:

```racket
#lang racket
(require redex/reduction-semantics)

(define-language λL
  (x ::= variable-not-otherwise-mentioned)
  (e ::= x (λ (x) e) (e e))
  #:binding-forms
  (λ (x) e #:refers-to x))

(define-metafunction λL
  free-vars : e -> (x ...)
  [(free-vars x)
   (x)]
  [(free-vars (λ (x) e))
   (- (free-vars e) (x))]
  [(free-vars (e_1 e_2))
   (∪ (free-vars e_1) (free-vars e_2))])

(define-metafunction λL
  - : (x ...) (x ...) -> (x ...)
  [(- any_ls1 any_ls2)
   ,(remf* (curryr member (term any_ls2)) (term any_ls1))])

(define-metafunction λL
  ∪ : (x ...) ... -> (x ...)
    [(∪ any_ls ...)
     ,(apply append (term (any_ls ...)))])

```

<details>
```
((emacs-version "25.1.1")
 (emacs-uptime "21 days, 23 hours, 40 minutes, 25 seconds")
 (system-type gnu/linux)
 (major-mode racket-mode)
 (racket--source-dir "/home/bluephoenix47/.emacs.d/elpa/racket-mode-20161101.1859/")
 (racket-racket-program "racket")
 (racket-raco-program "raco")
 (racket-memory-limit 2048)
 (racket-error-context medium)
 (racket-history-filter-regexp "\\`\\s *\\S ?\\S ?\\s *\\'")
 (racket-images-inline t)
 (racket-images-keep-last 100)
 (racket-images-system-viewer "display")
 (racket-pretty-print t)
 (racket-indent-curly-as-sequence t)
 (racket-indent-sequence-depth 0)
 (racket-pretty-lambda nil)
 (racket-smart-open-bracket-enable nil))
(enabled-minor-modes
 (TeX-PDF-mode)
 (async-bytecomp-package-mode)
 (auto-composition-mode)
 (auto-compression-mode)
 (auto-encryption-mode)
 (auto-fill-function)
 (auto-fill-mode)
 (auto-save-mode)
 (column-number-mode)
 (diff-auto-refine-mode)
 (electric-indent-mode)
 (evil-leader-mode)
 (evil-local-mode)
 (evil-mode)
 (evil-smartparens-mode)
 (evil-surround-mode)
 (file-name-shadow-mode)
 (font-lock-mode)
 (global-eldoc-mode)
 (global-evil-surround-mode)
 (global-font-lock-mode)
 (global-git-commit-mode)
 (global-undo-tree-mode)
 (hs-minor-mode)
 (line-number-mode)
 (menu-bar-mode)
 (override-global-mode)
 (save-place-mode)
 (server-buffer-clients)
 (server-mode)
 (shell-dirtrack-mode)
 (show-paren-mode)
 (smartparens-global-mode)
 (smartparens-mode)
 (tooltip-mode)
 (transient-mark-mode)
 (undo-tree-mode))
(disabled-minor-modes
 (LaTeX-math-mode)
 (TeX-Omega-mode)
 (TeX-interactive-mode)
 (TeX-source-correlate-mode)
 (abbrev-mode)
 (archive-subfile-mode)
 (blink-cursor-mode)
 (buffer-read-only)
 (compilation-in-progress)
 (compilation-minor-mode)
 (compilation-shell-minor-mode)
 (completion-in-region-mode)
 (cua-mode)
 (defining-kbd-macro)
 (desktop-save-mode)
 (diff-minor-mode)
 (dired-hide-details-mode)
 (display-time-mode)
 (doc-view-minor-mode)
 (eldoc-mode)
 (electric-layout-mode)
 (electric-quote-mode)
 (ert--current-run-stats)
 (evil-org-mode)
 (flyspell-mode)
 (gist-mode)
 (git-commit-mode)
 (global-evil-leader-mode)
 (global-prettify-symbols-mode)
 (global-visual-line-mode)
 (html-autoview-mode)
 (ido-everywhere)
 (image-minor-mode)
 (isearch-mode)
 (ispell-minor-mode)
 (jit-lock-debug-mode)
 (latex-electric-env-pair-mode)
 (mail-abbrevs-mode)
 (mml-mode)
 (next-error-follow-minor-mode)
 (org-capture-mode)
 (org-cdlatex-mode)
 (org-src-mode)
 (orgstruct-mode)
 (outline-minor-mode)
 (overwrite-mode)
 (paragraph-indent-minor-mode)
 (prettify-symbols-mode)
 (racket-check-syntax-mode)
 (rectangle-mark-mode)
 (sgml-electric-tag-pair-mode)
 (sh-electric-here-document-mode)
 (shell-command-with-editor-mode)
 (show-smartparens-global-mode)
 (show-smartparens-mode)
 (size-indication-mode)
 (smartparens-global-strict-mode)
 (smartparens-strict-mode)
 (smerge-mode)
 (tar-subfile-mode)
 (temp-buffer-resize-mode)
 (tool-bar-mode)
 (undo-tree-visualizer-selection-mode)
 (unify-8859-on-decoding-mode)
 (unify-8859-on-encoding-mode)
 (url-handler-mode)
 (use-hard-newlines)
 (vc-parent-buffer)
 (view-mode)
 (visible-mode)
 (visual-line-mode)
 (vlf-mode)
 (window-divider-mode)
 (with-editor-mode)
 (writegood-mode))
```
</details>

**Comment by greghendershott** [comment]
Just to confirm: I don't use Redex, but my guess (and trying briefly in DrRacket) is that only the last example lacks the desired indentation -- instead should be the following?

```
(define-metafunction λL
  ∪ : (x ...) ... -> (x ...)
  [(∪ any_ls ...)
   ,(apply append (term (any_ls ...)))])
```

**Comment by wilbowma** [comment]
Oh, yes I should have been explicit. The metafunction "-" is indented properly, but not that one named "\cup".

--

**Comment by greghendershott** [comment]
One difference is that if you put point on `∪` and do <kbd>C-u C-x =</kbd> it says

```
             position: 580 of 751 (77%), column: 4
            character: ∪ (displayed as ∪) (codepoint 8746, #o21052, #x222a)
    preferred charset: unicode (Unicode (ISO10646))
code point in charset: 0x222A
               script: symbol
               syntax: . 	which means: punctuation
```

The "punctuation" char-syntax seems to be confusing Emacs' `parse-partial-sexp`. If you replace `∪` with say `λ` which has "word" char-syntax, indent is as-expected.

Ugh. Indentation is my least-favorite aspect of doing racket-mode. There is a lot of `lisp-mode` functionality that e.g. `scheme-mode` inherits -- and it seems smart for me to inherit, not invent from scratch. On the other hand every time I look at the `lisp-mode` indent implementations, I get a massive headache.

Unfortunately I can't even think of a quick hack work-around for you (yet). Something like `(put 'define-metafunction 'racket-indent-function 1)` doesn't help.

**Comment by wilbowma** [comment]
Maybe I can redefine offending symbols as a word, as a workaround? Looking through emacs docs, there is a `put-char-code-property` (https://www.gnu.org/software/emacs/manual/html_node/elisp/Character-Properties.html), but I'm not sure thats related to the same interface.

**Comment by wilbowma** [comment]
Ahah, this did it: `(modify-syntax-entry ?∪ "w")` (via https://www.emacswiki.org/emacs/EmacsSyntaxTable)

**Comment by greghendershott** [comment]
That's a work-around for `∪`. I was hoping I could give you a work-around for _all_ such Unicode characters that Emacs thinks are "punctuation" -- without of course course trying to _enumerate_ all of them in a ginormous syntax table. :smile:

IIUC punctuation isn't even an interesting syntax class for lispy langs. So a plausible work-around might be to hack `char-syntax` to always return (say) "word" instead of "punctuation" (for racket-mode) But, even if that were possible, it's still a hack. The right place to fix this is _indentation_, not syntax classes.

Also there's #243. So, this corner is due for some work, already.

**Comment by greghendershott** [comment]
I decided to jump down the rabbit hole and overhaul the indentation.

I already have fixes for this and #243. I want to simplify things more, if possible, so I plan to keep working on that -- but if it takes too long I'll go ahead and push commit(s) for the fixes.
